dll/cpl/sysdm/environment.c

```diff
@@ -45,97 +145,423 @@ GetSelectedListViewItem(HWND hwndListView)
+    SendDlgItemMessage(hwndDlg,
+                       IDC_LIST_VARIABLE_VALUE,
+                       LB_GETITEMRECT,
+                       DlgData->dwSelectedValueIndex,
+                       (LPARAM)&rItem);
+
+    /* Get string from item */
+    SendDlgItemMessage(hwndDlg,
+                       IDC_LIST_VARIABLE_VALUE,
+                       LB_GETTEXT,
+                       DlgData->dwSelectedValueIndex,
+                       (LPARAM)szStr);
+
+    DlgData->hEditBox = CreateWindow(_T("EDIT"),
+                                     szStr,
+                                     WS_VISIBLE | WS_CHILD | WS_BORDER |
+                                     ES_LEFT | ES_AUTOHSCROLL,
+                                     rItem.left, rItem.top,
+                                     rItem.right, rItem.bottom - rItem.top,
+                                     GetDlgItem(hwndDlg, IDC_LIST_VARIABLE_VALUE),
+                                     NULL,
+                                     hApplet,
+                                     NULL);
+
+    hFont = (HFONT)SendDlgItemMessage(hwndDlg, IDC_LIST_VARIABLE_VALUE, WM_GETFONT, 0, 0);
+    SendMessage(DlgData->hEditBox, WM_SETFONT, (WPARAM)hFont, TRUE);
+    SendMessage(DlgData->hEditBox, EM_SETSEL, 0, -1);
+    SetFocus(DlgData->hEditBox);
+    DlgData->dwTextEditedValueIndex = DlgData->dwSelectedValueIndex;
+}
+
+static VOID
+ItemEditEnds(HWND hwndDlg,
+             PDIALOG_DATA DlgData)
+{
+    TCHAR szStr[MAX_PATH] = _T("");
+    TCHAR szStr2[MAX_PATH] = _T("");
+
+    SendMessage(DlgData->hEditBox, WM_GETTEXT, MAX_PATH, (LPARAM)szStr);
+    SendDlgItemMessage(hwndDlg,
+                       IDC_LIST_VARIABLE_VALUE,
+                       LB_GETTEXT,
+                       DlgData->dwTextEditedValueIndex,
+                       (LPARAM)szStr2);
+    SendDlgItemMessage(hwndDlg,
+                       IDC_LIST_VARIABLE_VALUE,
+                       LB_DELETESTRING,
+                       DlgData->dwTextEditedValueIndex, 0);
+
+    if (wcscmp(szStr, _T("")) != 0 || wcscmp(szStr, szStr2) != 0)
+    {
+        SendDlgItemMessage(hwndDlg,
+                           IDC_LIST_VARIABLE_VALUE,
+                           LB_INSERTSTRING,
+                           DlgData->dwTextEditedValueIndex, (LPARAM)szStr);
+    }
+    else
+    {
+        SendDlgItemMessage(hwndDlg,
+                           IDC_LIST_VARIABLE_VALUE,
+                           LB_SETCURSEL,
+                           DlgData->dwTextEditedValueIndex - 1, 0);
+
+        DlgData->dwSelectedValueIndex = DlgData->dwTextEditedValueIndex - 1;
+    }
+
+    DestroyWindow(DlgData->hEditBox);
+    DlgData->dwTextEditedValueIndex = -1;
+}
+
+static VOID
+MoveListItem(HWND hwndDlg,
+             PDIALOG_DATA DlgData,
+             BOOL bMoveUp)
+{
+    TCHAR szDest[MAX_PATH];
+    DWORD dwSrcIndex, dwDestIndex, dwTotalIndex;
+    
+    dwTotalIndex = SendDlgItemMessage(hwndDlg, IDC_LIST_VARIABLE_VALUE, LB_GETCOUNT, 0, 0) - 1;
```

Comment:
Can you test this MoveListItem() when your path env-var contains only one item? And no items at all?

Reply:
I did, no problems I faced.